src/tools/database/sql.ts

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright 2024 IBM Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Tool,
+  ToolInput,
+  BaseToolOptions,
+  BaseToolRunOptions,
+  JSONToolOutput,
+} from "@/tools/base.js";
+import { z } from "zod";
+import { getMetadata } from "@/tools/database/metadata.js";
+import { isSupported, connectSql } from "@/tools/database/connection.js";
+
+type ToolOptions = { provider: string } & BaseToolOptions;
+type ToolRunOptions = BaseToolRunOptions;
+
+export class SQLTool extends Tool<JSONToolOutput<any>, ToolOptions, ToolRunOptions> {
+  name = "SQLTool";
+
+  description =
+    "Converts natural language to SQL query and executes it using the provided tables schema.";
+
+  inputSchema() {
+    return z.object({
+      query: z.string({ description: "The SQL query to be executed." }).min(1),
+    });
+  }
+
+  static {
+    this.register();
+  }
+
+  protected async _run(
+    { query }: ToolInput<this>,
+    _options?: ToolRunOptions,
+  ): Promise<JSONToolOutput<any>> {
+    const provider = this.options.provider;
+
+    if (!isSupported(provider)) {
+      throw new Error(`Unsupported database provider: ${provider}`);
```

Comment:
always throw an instance of the `ToolError` class or a class that extends it

Reply:
also you should do this check in the constructor to throw as soon as possible

Reply:
not needed after strong typing of provider